src/wasm/literal.cpp

```diff
@@ -837,16 +837,61 @@ Literal Literal::demote() const {
   return Literal(float(getf64()));
 }
 
+// Wasm has nondeterministic rules for NaN propagation in some operations. For
+// example. f32.neg is deterministic and just flips the sign, even of a NaN, but
+// f32.add is nondeterministic, and if one or more of the inputs is a NaN, then
+//
+//  * if all NaNs are canonical NaNs, the output is some arbitrary canonical NaN
+//  * otherwise the output is some arbitrary arithmetic NaN
+//
+// (canonical = NaN payload is 1000..000; arithmetic: 1???..???, that is, the
+// high bit is 1 and all others can be 0 or 1)
+//
+// For many things we don't need to care, and can just do a normal C++ add for
+// an f32.add, for example - the wasm rules are specified so that things like
+// that just work (in order for such math to be fast). However, for our
+// optimizer, it is useful to "standardize" NaNs when there is nondeterminism.
+// That is, when there are multiple valid outputs, it's nice to emit the same
+// one consistently, so that it doesn't look like the optimization changed
+// something. In other words, if the valid output of an expression is a set of
+// valid NaNs, and after optimization the output is still that same set, then
+// the optimization is valid. And if the interpreter picks the same NaN in both
+// cases from that identical set then nothing looks wrong to the fuzzer.
+//
+// Note that this function must receive both inputs to the operation, so that it
+// can tell if a NaN existed beforehand (if a NaN was generated just now, as in
+// 0/0, then we should not do any standardization).
+template<typename T> static Literal standardizeNaN(T left, T right, T result) {
+  if (!std::isnan(left) && !std::isnan(right)) {
```

Comment:
Is this necessary? It seems like you can check whether the just the `result` is a NaN value. You mention this above: "if a NaN was generated just now, as in 0/0, then we should not do any standardization", but I don't see why that would be true. It's still possible for that operation to produce a non-deterministic NaN.

Reply:
Thanks, I see you're right, I missed this in the spec: `including the case that there are no NaN inputs`. Updated.